### ISSUE TITLE
cargo / crates.io is not immutable - potential risk of left-pad incident

### ISSUE THREAD

**Issue by fschutt** [rust-lang/cargo#5263]
This is an issue that I'm a bit concerned about: When yanking a crate, cargo does not check if that crate is dependent upon by other crates. Therefore, a [left-pad incident](http://blog.npmjs.org/post/141577284765/kik-left-pad-and-npm) can currently happen. If someone would decide to yank `libc` right now, for example, all hell would break loose.

Example: https://crates.io/crates/libfoo-test

Here we have libfoo, which depends on [libbaz](https://crates.io/crates/libbaz-test) version 0.1.0. I was able to yank libbaz 0.1.0, thereby breaking libfoo in the process. What you have to do is to publish a new version of libbaz, then yank the (version - 1) crate. For some reason, cargo neither does semver-checks (it could, for example, see that version 0.1.0 and 0.2.0 are identical) nor does it allow libfoo access to libbaz (in a "you know my sha1, so I'll give you access, even though the crate is yanked" way).

If you try to depend on libfoo now, all you get is:

```
    Updating registry `https://github.com/rust-lang/crates.io-index`
error: no matching version `^0.1.0` found for package `libbaz-test`
location searched: registry `https://github.com/rust-lang/crates.io-index`
versions found: 0.2.0
required by package `libfoo-test v0.1.0`
    ... which is depended on by `dependency-test v0.1.0 (file:///home/felix/Development/dependency-test)`
```

This is dangerous because now the maintainer of libfoo has to be notified to manually update his package to upgrade to libbaz 0.2.0. If libbaz has lots of dependents, this can break a lot of builds. Yes, the maintainer can upgrade to the new version, but the damage is done at first. This violates the "immutability" of cargo - once a package (in this case libfo) is uploaded, it should always build. Is this the intentional design? If yes, why?

**Comment by sfackler** [comment]
> nor does it allow libfoo access to libbaz (in a "you know my sha1, so I'll give you access, even though the crate is yanked" way)

It does allow this. An existing Cargo.lock will still be able to use a yanked version.

**Comment by matklad** [comment]
Yep, as @sfackler says, cargo does not really suffers from left-pad problem, because yank **does not remove** code, and, if you have Cargo.lock around, you still will be able to use the yanked version.

The relevant docs are here: https://doc.rust-lang.org/cargo/reference/publishing.html#managing-a-cratesio-based-crate

**Comment by fschutt** [comment]
Great. This just broke my code **yet again**. The problem is that **libraries do not submit their lock files to crates.io because cargo puts them in the gitignore by default**. dwrote decided to unpublish itself while I was asleep, and now I got a report that someone can't install my library or run any examples.

**Comment by Eh2406** [comment]
cc https://github.com/rust-lang/cargo/pull/5967

**Comment by fschutt** [comment]
Well, the real "problem" I have is that if there is a dependency that has a lock file, the cargo / crates.io seems to ignore that lock file. For example, if I submit a library to crates.io with a Cargo.lock in it, then I'd expect cargo to recognize that and use the hashes in that lock file, so that even if a dependent crate gets yanked, my crate still works.
